app/models/authentication.rb

```diff
@@ -5,10 +5,19 @@ class Authentication < ApplicationRecord
   encrypt_column :password
 
   belongs_to :resource, :polymorphic => true
+  belongs_to :source
 
   has_many :application_authentications, :dependent => :destroy
   has_many :applications, :through => :application_authentications
 
   attribute :availability_status, :string
   validates :availability_status, :inclusion => { :in => %w[available unavailable] }, :allow_nil => true
+
+  before_validation :set_source
+
+  private
+
+  def set_source
+    self.source_id = resource.source_id
```

Comment:
resource can be nil

Reply:
Well, the fun thing is that it won't allow it to be created unless the source exists, so `nil` won't work. I'll add `:optional => true` to the `belongs_to` statement to allow a blank source_id.